### PR TITLE
[wip] QAT ConvBN: remove explicit folding and use BN instead

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38478 [wip] QAT ConvBN: remove explicit folding and use BN instead**

Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Training MobileNetV2 on Imagenet-1k with QAT - accuracy curves match exactly before and after this diff

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21603230](https://our.internmc.facebook.com/intern/diff/D21603230)

### CODE REVIEW

File: test/quantization/test_qat_module.py [hunk 1]
```diff
@@ -82,6 +270,7 @@ def test_conv_bn_relu(
 
             cls = ConvBnReLU2d if use_relu else ConvBn2d
             qat_op = cls(
+                copy.deepcopy(bn_op),
```
Comment (jerryzh168) [inline test/quantization/test_qat_module.py:273]: can you move bn after the conv2d parameters

Comment (vkuzo) [reply test/quantization/test_qat_module.py:273]: one limitation is that some of the conv params are optional, and all required params have to come before optional params.  I thought this was worth changing the arg order (so we can make bn required), but flexible - lmk.

File: torch/nn/intrinsic/qat/modules/conv_fused.py [hunk 2]
```diff
@@ -49,83 +44,45 @@ def __init__(self,
...
-
-        # we use running statistics from the previous batch, so this is an
-        # approximation of the approach mentioned in the whitepaper, but we only
-        # need to do one convolution in this case instead of two
-        running_std = torch.sqrt(self.running_var + self.eps)
-        scale_factor = self.gamma / running_std
+        running_std = torch.sqrt(self.bn.running_var + self.bn.eps)
+        scale_factor = self.bn.weight / running_std
         scaled_weight = self.weight * scale_factor.reshape([-1, 1, 1, 1])
+        # this does not include the conv bias
```
Comment (raghuramank100) [inline torch/nn/intrinsic/qat/modules/conv_fused.py:80]: Defining scaled_weight as:
```
 scaled_weight = self.weight_fake_quant(self.weight * scale_factor.reshape([-1, 1, 1, 1]))/scale_factor.reshape([1, -1, ,1,1])
```
simplifies the code to:
```
 conv = self._conv_forward(input, scaled_weight)
```

**Review by raghuramank100 — APPROVED** [review]
Looks good to me!

File: test/quantization/test_qat_module.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_qat_module.py:273]: OK

File: test/quantization/test_qat_module.py [hunk 3]
```diff
@@ -3,19 +3,207 @@
...
-from torch.nn import Conv2d, BatchNorm2d, ReLU
+import torch.nn as nn
+from torch.nn import Conv2d, BatchNorm2d, ReLU, init
 from torch.nn.intrinsic.qat import ConvBn2d, ConvBnReLU2d
+from torch.nn.modules.utils import _pair
 from torch.quantization.qconfig import default_qat_qconfig
 import torch.backends.mkldnn
 from torch.testing._internal.common_utils import TestCase
-from hypothesis import given
+from hypothesis import given, seed
```
Comment (jerryzh168) [inline test/quantization/test_qat_module.py:15]: lint

**Review by jerryzh168 — APPROVED** [review]
Looks good

File: test/quantization/test_qat_module.py (hunk 1 above)
Comment (vkuzo) [reply test/quantization/test_qat_module.py:274]: actually, I take it back, passing the module instead of params leads to other problems.  Will switch back to params and move it after conv
